### ISSUE TITLE
Printing complex/large variance-covariance matrices

### ISSUE THREAD

**Issue by kaskr** [glmmTMB/glmmTMB#196]
Try this:

```
library(glmmTMB)
example(glmmTMB)
sleepstudy$row <- factor(1:180)
fm <- glmmTMB(Reaction ~ ar1(row + 0 | Subject), sleepstudy)
```

Printing the model takes a very long time and prints too much (and for some reason emacs doesn't seem to like scrolling the output) :

```
system.time(print(fm))

   user  system elapsed 
  0.064   0.064  14.796 
```

The problem is `glmmTMB:::print.VarCorr.glmmTMB`. When the corr matrices do not fit the display I'd prefer by default either

1. Printing just the head cols/rows **or**
2. Printing both head and tail cols/rows

Perhaps this is just a matter of passing appropriate arguments to `lme4:::formatVC`...
Any thoughts ?

**Comment by bbolker** [comment]
In the console (linux) it's not so bad -- 0.144 user time, 0.472 elapsed.  (But it took emacs 70 seconds to print on my system.) Regardless of the timing, though, the print format is ridiculous. I think complex variance-covariance matrices should know something about their own structure and how to print themselves -- this was done nicely in nlme, which had a range of matrix types with their own print methods.

**Comment by mmaechler** [comment]
I agree 
- that the current print method is ridiculous as soon as the matrices are non small
- ideally should change this in lme4:::formatVC or another "place" that's common to the two packages.
  IIRC, we don't by default print the VC matrix in summary(<lme4>), do we?

**Comment by bbolker** [comment]
I pushed a preliminary version of this at  cf9237d . I copied `formatVC` from `lme4` and hacked it (but it's backward compatible; if we do want to share it I would propose we eventually put it back in `lme4` and export it from there). I'm not sure about the structure/design: ideally it would make sense for the *individual* components of variance-covariance objects to have classes ... for now, though, I just wanted to get something working.  Right now it looks like this (use the `print_fancy_vcov` branch ...)

```
library(glmmTMB)
data(sleepstudy,package="lme4")
sleepstudy$row <- factor(1:180)
fm <- glmmTMB(Reaction ~ ar1(row + 0 | Subject), sleepstudy)
VarCorr(fm)
## Conditional model:
##   Groups   Name Std.Dev.   Corr       
##   Subject  row1 6.1192e+01 0.853 (ar1)
##  Residual          3.4087e-07     
```

Comments welcome.

**Comment by bbolker** [comment]
I've been poking into the covariance structure stuff a bit, and have the following thoughts going forward:

- the homogeneous-variance models (ar1, ou, exp, gau, mat) are the ones that are most pressing, as it's most likely that the variance-covariance matrices will be large *and* not contain much information. For heterogeneous-variance models (us, diag, cs, toep), even if we have a small number of covariance/correlation values (e.g. cs, diag), we still have to print out all those rows. (Arguably, we should still truncate the number of rows - but it's also the case as the number of variance-covariance parameters increases that the model becomes more likely to become silly/overparameterized anyway ...)
- it might be a good idea to allow homogeneous-variance versions of `diag`, `cs`, `us`. `MCMCglmm` has `us`, `idv` (= our `diag`), `idh` (diag with homogeneous variance), `cor*` (correlation matrices), and `ante*` ("antedependence" - autocorrelation?)

**Comment by kaskr** [comment]
I agree that we should add the homogeneous strucures as well. This can be handled on the R-side by collecting / fixing parameters in existing covstructs. I'd also like to add general AR(k) and ante dependence.

Eventually, there will be a large number of options. Having a print method for them all does not seem feasible ?

Perhaps it is time to consider alternative ways to report covariance matrices from C++ ? Our current format ( an `sd` vector and a `corr` matrix) is both inefficient for large models and inappropriate for printing. Some simple alternatives:

1. Report the correlation matrix as a `dgTMatrix` with only the relevant elements set. E.g for `ar1` we would only need a subset of the first row whereas all elements should be set for `us`.
2. Report the correlation indirectly in terms of a named parameter vector e.g. something like `corr_range` and `corr_smoothness` for the Matern correlation.
3. A combination of 1 and 2.

It should be easier to work out a generic print method once we get the output format right ?

**Comment by mebrooks** [comment]
Done with pull #243

**Comment by kaskr** [comment]
Jarrod Hadfield is asking where to find the AR1 correlation parameter in the glmmTMB output.
It seems the print method, originally pulled in from the `print_fancy_cov` branch, is now broken.
This is what now comes out of @bbolker's original example that used to work (see previous example in this issue):

```
library(glmmTMB)
data(sleepstudy,package="lme4")
sleepstudy$row <- factor(1:180)
fm <- glmmTMB(Reaction ~ ar1(row + 0 | Subject), sleepstudy)
VarCorr(fm) 
## Conditional model:
##  Groups   Name Std.Dev.   Corr  
##  Subject  row1 6.1192e+01  (ar1)
##  Residual      3.4086e-07       
```

The correlation parameter has disappeared ! Any idea what could have happened ?

**Comment by bbolker** [comment]
I'll see what I can do (not before tonight though ...)

**Comment by bbolker** [comment]
should be fixed in  c105d9f ... puzzled as to why this worked previously, but I don't want to do the necessary archaeology to find out. Maybe it didn't??  Added a few comments to the formatVC code (which was inherited from lme4 and had already been modified/commented on to be just a little more scrutable).
